airflow/models/dag.py

```diff
@@ -2435,6 +2436,56 @@ def cli(self):
         args = parser.parse_args()
         args.func(args, self)
 
+    @provide_session
+    def test(self, execution_date: datetime | None = None, run_conf=None, session=None):
+        """Execute one single DagRun for a given DAG and execution date, using the DebugExecutor."""
+
+        def add_logger_if_needed(ti: TaskInstance):
+            """
+            Add a formatted logger to the taskinstance so all logs are surfaced to the command line instead
+            of into a task file. Since this is a local test run, it is much better for the user to see logs
+            in the command line, rather than needing to search for a log file.
+            Args:
+                ti: The taskinstance that will receive a logger
+
+            """
+            format = logging.Formatter("[%(asctime)s] {%(filename)s:%(lineno)d} %(levelname)s - %(message)s")
+            handler = logging.StreamHandler(sys.stdout)
+            handler.level = logging.INFO
+            handler.setFormatter(format)
+            handler_classes = [h.__class__.__name__ for h in ti.log.handlers]
+            if "StreamHandler" not in handler_classes:  # only add log handler once
+                ti.log.addHandler(handler)
+
+        execution_date = execution_date or timezone.utcnow()
+        self.clear(
+            start_date=execution_date,
+            end_date=execution_date,
+            dag_run_state=False,  # type: ignore
+            session=session,
+        )
+
+        dr: DagRun = _get_or_create_dagrun(
+            dag=self,
+            start_date=execution_date,
+            execution_date=execution_date,
+            run_id=DagRun.generate_run_id(DagRunType.MANUAL, execution_date),
+            session=session,
+            conf=run_conf,
+        )
+
+        tasks = self.task_dict
+        log.info("starting dagrun")
+        # Instead of starting a scheduler, we run the minimal loop possible to check
+        # for task readiness and dependency management. This is notably faster
+        # than creating a BackfillJob and allows us to surface logs to the user
+        while dr.state == State.RUNNING:
+            schedulable_tis, _ = dr.update_state(session=session)
+            for ti in schedulable_tis:
+                add_logger_if_needed(ti)
+                ti.task = tasks[ti.task_id]
+                _run_task(ti, session=session)
```

Comment:
I wonder if it’s a good idea to extract the logic in `SchedulerJob._schedule_dag_run()` (which this loop replicates).

Reply:
I feel somewhat hesitant on that as the `_schedule_dag_run` seems to be much more interested in scheduling than it is in actually running. We'd need to cut at some potentially awkward places to ensure that tasks are run instead of just scheduled to run. I also think that would involve touching some very core code in the scheduler unnecessarily. @ashb WDYT?